Add ControllsWidget

JannisPetschenka opened the pull request:
Edited the widget to allow a more generic config :)

ErikReider commented:
Pressed the wrong button, my bad...

JannisPetschenka commented on README.md:
```diff
@@ -49,6 +49,91 @@ These widgets can be customized, added, removed and even reordered
+        {
+          "label": "Performance",
+          "command": "powerprofilesctl set performance"
+        },
+        {
+          "label": "Balanced",
+          "command": "powerprofilesctl set balanced"
+        },
+        {
+          "label": "Power-saver",
+          "command": "powerprofilesctl set power-saver"
+        }
+      ]
+    },
+    "controll-buttons": {
+      "type": "buttons",
+      "actions": [
+        {
+          "label": "直",
+          "command": "~/.config/rofi/rofi-wifi-menu.sh"
+        },
+        {
+          "label": "",
+          "command": "~/.config/rofi/rofi-bluetooth"
+        }
+      ]
+    },
+    "topbar-buttons": {
+      "type": "buttons",
+      "position": "topbar-right",
+      "actions": [
+        {
+          "label": "",
+          "command": "...",
+        }
+      ]
+    }
+  }
+  ...
+}
```
Hmm, ok. Since the menus (and also the buttons if you choose `"position": "topbar-right"`) will be placed into the same `topbar_container` I will have to edit the `controlCenter.ui`. I'll check it out :)

ErikReider replied:
Oh, there's positioning! Didn't notice... 😅

I wonder if there's a better way of doing this than my suggestion above?

ErikReider replied:
Maybe something like this?

```json
"widgets": [
  "menubar#label",
  "buttons",
  "buttons#AnotherButton123",
  "title",
  "dnd",
  "mpris",
  "notifications"
],
"widget-config": {
  "menubar#label": {
    "menus": [
      {
        "label": "Generic menu item",
        "position": "right",
        "actions": [ 
          {
            "label": " Reboot",
            "command": "systemctl reboot"
          },
          ...
        ]
      },
      {
        "label": "Another menu config",
        "position": "left",
        "actions": [ 
          {
            "label": " Reboot",
            "command": "systemctl reboot"
          },
          ...
        ]
      },
    ]
  },
  "buttons": {
    "actions": [
      {
        "label": "直",
        "command": "~/.config/rofi/rofi-wifi-menu.sh"
      },
      ...
    ]
  },
  "buttons#AnotherButton123": {
    "actions": [
      {
        "label": "直",
        "command": "~/.config/rofi/rofi-wifi-menu.sh"
      },
      ...
    ]
  },
}
```

Would that work?

JannisPetschenka replied:
Jup, that shoud work. I would add an additional type to the menu so you could add buttons, which would not open a dropdown (e.g. the screenshot button in my config). Will change it :)

ErikReider replied:
Cool! Thanks for your patience 😅

JannisPetschenka replied:
Done. I might have to add some more checks when parsing the config, but then it should be ready :)

ErikReider commented:
What's the difference between 
```json
"buttons": {
  "actions": [
    {
      "label": "直",
      "command": "~/script.sh"
    },
    {
      "label": "",
      "command": "~/script.sh"
    }
  ]
},
```
and
```json
"menubar": {
  "power-buttons": {
    "label": "", 
    "position": "right",
    "actions": [ 
      {
        "label": " Reboot",
        "command": "~/script.sh"
      },
      {
        "label": " Shut down",
        "command": "~/script.sh"
      }
    ]
  },
}
```

Isn't "buttons" essentially just a menubar? Would merging them make sense due to being able to have buttons in the menubar? Am I missing something?

JannisPetschenka commented:
> What's the difference between
> 
> ```json
> "buttons": {
>   "actions": [
>     {
>       "label": "直",
>       "command": "~/script.sh"
>     },
>     {
>       "label": "",
>       "command": "~/script.sh"
>     }
>   ]
> },
> ```
> 
> and
> 
> ```json
> "menubar": {
>   "topbar-buttons": {
>     "type": "buttons",
>     "position": "right",
>     "actions": [
>       {
>         "label": "直",
>         "command": "~/script.sh"
>       }
>     ]
>   },
> }
> ```
> 
> Isn't "buttons" essentially just a menubar? Would merging them make sense due to being able to have buttons in the menubar? Am I missing something?

It's meant to be something different, probably my bad for using the same command and label everywhere :) 

The buttons-widget is a FlowBox so it's meant to display a grid of buttons. The button type in the menu-widget however is for displaying buttons in the topbar, which do not open a drop-down (e.g. a screenshot button), so it's not meant to display more than one line of buttons.

This example should show what I mean.
![image](https://user-images.githubusercontent.com/78504175/214917128-63e2b7d5-c014-49ba-81ae-d85b72302eb4.png)

ErikReider commented on README.md:
```diff
@@ -49,6 +49,72 @@ These widgets can be customized, added, removed and even reordered
+{
+  ...
+  "menubar#label": {
+    "power-buttons": {
```
We should probably use `button|menu#AnyName` instead so that we can skip the `type` in the buttons one below

ErikReider commented on src/controlCenter/widgets/buttons/buttons.vala:
```diff
@@ -0,0 +1,43 @@
+                if (a != null) actions = parse_actions (a);
+            }
+
+            container = new Gtk.FlowBox ();
```
Variable can be local

ErikReider commented on src/controlCenter/widgets/buttons/buttons.vala:
```diff
@@ -0,0 +1,43 @@
+                container.insert (b, -1);
+            }
+
+            pack_start (container, true, true, 0);
```
Move this up to the container creation

ErikReider commented:
Haven't really had time to look at the `menubar.vala` file yet. Should have more time tomorrow :)

ErikReider commented:
Sorry for the delay. I've had a lot on my plate the past 2 weeks. I should have time within the next few days :)

ErikReider commented on src/controlCenter/widgets/menubar/menubar.vala:
```diff
@@ -0,0 +1,163 @@
+
+                show_button.clicked.connect (() => {
+                    if (obj.hidden) {
+                        menus.foreach (m => m.hide ());
```
Lets add a nullable menu box into each `ConfigObject` and remove the `menus` list. There's a bug when you're trying to open multiple menus while one's still open which gets solved by this:

```diff
diff --git a/src/controlCenter/widgets/menubar/menubar.vala b/src/controlCenter/widgets/menubar/menubar.vala
index c2b6511..c9e56d8 100644
--- a/src/controlCenter/widgets/menubar/menubar.vala
+++ b/src/controlCenter/widgets/menubar/menubar.vala
@@ -8,7 +8,7 @@ namespace SwayNotificationCenter.Widgets {
         string ? label;
         string ? position;
         Action[] actions;
-        bool hidden;
+        Gtk.Box? menu;
     }
 
     public struct Action {
@@ -25,9 +25,8 @@ namespace SwayNotificationCenter.Widgets {
 
         Gtk.Box menus_container;
         Gtk.Box topbar_container;
-        List<Gtk.Box> menus;
 
-        ConfigObject[] menu_objects;
+        List<ConfigObject?> menu_objects;
 
         public Menubar (string suffix, SwayncDaemon swaync_daemon, NotiDaemon noti_daemon) {
             base (suffix, swaync_daemon, noti_daemon);
@@ -44,17 +43,20 @@ namespace SwayNotificationCenter.Widgets {
 
             menus_container.add (topbar_container);
 
-            foreach (var obj in menu_objects) {
-                add_menu (obj);
+            for (int i = 0; i < menu_objects.length (); i++) {
+                unowned ConfigObject? obj = menu_objects.nth_data (i);
+                add_menu (ref obj);
             }
 
             pack_start (menus_container, true, true, 0);
             show_all ();
 
-            menus.foreach (m => m.hide ());
+            foreach (var obj in menu_objects) {
+                obj.menu?.hide ();
+            }
         }
 
-        void add_menu (ConfigObject obj) {
+        void add_menu (ref unowned ConfigObject? obj) {
             if (obj.type == "buttons") {
                 Gtk.Box container = new Gtk.Box (Gtk.Orientation.HORIZONTAL, 0);
                 container.get_style_context ().add_class (obj.name);
@@ -81,18 +83,14 @@ namespace SwayNotificationCenter.Widgets {
 
                 Gtk.Box menu = new Gtk.Box (Gtk.Orientation.VERTICAL, 0);
                 menu.get_style_context ().add_class (obj.name);
-                menus.append (menu);
-                obj.hidden = true;
+                obj.menu = menu;
 
                 show_button.clicked.connect (() => {
-                    if (obj.hidden) {
-                        menus.foreach (m => m.hide ());
-                        menu.show ();
-                        obj.hidden = !obj.hidden;
-                    } else {
-                        menu.hide ();
-                        obj.hidden = !obj.hidden;
+                    bool visible = !menu.visible;
+                    foreach (var o in menu_objects) {
+                        o.menu?.set_visible (false);
                     }
+                    menu.set_visible (visible);
                 });
 
                 foreach (var a in obj.actions) {
@@ -120,7 +118,7 @@ namespace SwayNotificationCenter.Widgets {
         protected void parse_config_objects (Json.Object config) {
             var elements = config.get_members ();
 
-            menu_objects = new ConfigObject[elements.length ()];
+            menu_objects = new List<ConfigObject?> ();
             for (int i = 0; i < elements.length (); i++) {
                 string e = elements.nth_data (i);
                 Json.Object ? obj = config.get_object_member (e);
@@ -148,14 +146,14 @@ namespace SwayNotificationCenter.Widgets {
                     }
 
                     Action[] actions_list = parse_actions (actions);
-                    menu_objects[i] = ConfigObject () {
+                    menu_objects.append (ConfigObject () {
                         name = name,
                         type = type,
                         label = label,
                         position = pos,
                         actions = actions_list,
-                        hidden = true
-                    };
+                        menu = null,
+                    });
                 }
             }
         }
```

I also changed the array into a GLib List due to some indexes not being filled if that specific menu doesn't get added (due to config error)

ErikReider commented on src/controlCenter/widgets/menubar/menubar.vala:
```diff
@@ -0,0 +1,163 @@
+                    menu.pack_start (b, true, true, 0);
+                }
+
+                if (obj.position == "right") {
+                    topbar_container.pack_end (show_button, false, false, 0);
+                } else if (obj.position == "left") {
+                    topbar_container.pack_start (show_button, false, false, 0);
+                } else {
+                    debug ("Invalid position for menu item in config");
+                }
```
Also a switch with enum

ErikReider reviewed (changes requested):
Finally found some time to review this. Thanks for your patience! :)

JannisPetschenka commented on README.md:
```diff
@@ -49,6 +49,82 @@ These widgets can be customized, added, removed and even reordered
+        "position": "left",
+        "actions": [ 
+          {
+            "label": "Performance",
+            "command": "powerprofilesctl set performance"
+          },
+          ...
+        ]
+      },
+      "buttons#topbar-buttons": {
+        "position": "right",
+        "actions": [
+          {
+            "label": "",
+            "command": "grim"
+          },
+        ]
+      }
+    },
+    "buttons": {
+      "actions": [
+        {
+          "label": "直",
+          "command": "~/.config/rofi/rofi-wifi-menu.sh"
+        },
+        {
+          "label": "",
+          "command": "~/.config/rofi/rofi-bluetooth"
+        },
+        ...
+      ]
+    },
+    "buttons#Another": {
+      "actions": [
+        ...
+      ]
+    }
+    ...
+  }
+}
```
Updated the `configSchema.json` not quite sure about the `ref` but I think it should be correct :)

ErikReider commented on src/controlCenter/widgets/buttons/buttons.vala:
```diff
@@ -0,0 +1,38 @@
+                if (a != null) actions = parse_actions (a);
+            }
+
+            Gtk.FlowBox container = new Gtk.FlowBox ();
```
Set the FlowBox selection mode to none. Fixes the blue FlowBoxChild focus coloring:

![image](https://user-images.githubusercontent.com/35975961/217660962-c474293d-63d6-43ef-b55f-dee1af49c9e1.png)

ErikReider reviewed (changes requested):
![image](https://user-images.githubusercontent.com/35975961/217660525-4d6d4e6f-82b5-4063-82a8-931beec09e4c.png)

Seems like the default CSS is a bit broken.

Other than this, I'll merge it :)

JannisPetschenka commented:
Updated the man(5) and the other issues :) Also realized, that the difference between the buttons widget and the buttons object in the menubar widget was getting kinda confusing.. So I changed the name of the buttons widget to buttons-grid :)

ErikReider reviewed (approved):
LGTM! Thanks for your patience! ❤️
